### ISSUE TITLE
invalid-return-type instead of invalid-argument-type with Self classmethod argument.

### ISSUE THREAD

**Issue by anatolegosset** [astral-sh/ty#4036]
### Summary

When running ty check on the following code:

```py
from typing import Self


class MyClass:
    def __init__(self, value: int):
        self.value = value

    @classmethod
    def construct(cls, from_value: int | Self) -> Self:
        if isinstance(from_value, cls):
            return from_value
        if isinstance(from_value, int):
            return cls(value=from_value)
        raise TypeError(
            f"Cannot construct MyClass from value of type {type(from_value)}",
        )


def my_function(value: int | None) -> MyClass:
    my_instance = MyClass.construct(value)
    return my_instance
```

ty indicates a `invalid-return-type` error on line 21 `return my_instance`, saying that the inferred type of `my_instance` is `MyClass | None`, incompatible with the `MyClass` explicit typing of `my_function`. This seems incorrect to me, since `MyClass.construct` can only return a `MyClass` instance. Instead, I would expect ty to indicate an `invalid-argument-type` error on line 20, specifically when calling `MyClass.construct` with `value: int | None`, which is incompaible with the `from_value: int | Self` declaration. 

To be clear, I do not expect ty to validate this code, but I expected a different error.

### Version

ty 0.0.61

**Comment by AlexWaygood** [comment]
Thank you! Very interesting -- here's a slightly more minimal repro:

```py
from typing import Self

class MyClass:
    @classmethod
    def not_broken(cls, x: Self) -> Self:
        return x
    
    @classmethod
    def broken(cls, x: Self | int) -> Self:
        return cls()

MyClass.not_broken(None)  # correctly fails
MyClass.broken(None)  # correctly fails

def f(x: int | None):
    MyClass.not_broken(x)  # correctly fails
    reveal_type(MyClass.broken(x))  # incorrectly accepted, reveals `MyClass | None`
```

**Comment by AlexWaygood** [comment]
And the same issue exists if it's an instance method:

```py
from typing import Self

class MyClass:
    def not_broken(self, x: Self) -> Self:
        return x
    
    def broken(self, x: Self | int) -> Self:
        return self

MyClass().not_broken(None)  # correctly fails
MyClass().broken(None)  # correctly fails

def f(x: int | None):
    MyClass().not_broken(x)  # correctly fails
    reveal_type(MyClass().broken(x))  # incorrectly accepted, reveals `MyClass | None`
```

**Comment by AlexWaygood** [comment]
Ah, and here's a more minimal repro yet that demonstrates that this isn't anything specific to `Self` type variables -- it's a general issue with type-variable solving:

```py
class MyClass: ...

def broken[T: MyClass](self: T, x: T | int) -> T:
    return self

def f(x: int | None):
    broken(MyClass(), x)  # no error here, but there should be -- `None` doesn't satisfy `int` or the upper bound of `T`
```

https://play.ty.dev/f3fc0240-3f65-41ab-9182-360960696b2c
